bin/tx-prover/src/proxy/mod.rs

```diff
@@ -0,0 +1,255 @@
+            http_peer.get_mut_peer_options().ok_or(Error::new(ErrorType::InternalError))?;
+
+        // Timeout settings
+        peer_opts.total_connection_timeout = TIMEOUT_SECS;
+        peer_opts.connection_timeout = CONNECTION_TIMEOUT_SECS;
+        peer_opts.read_timeout = TIMEOUT_SECS;
+        peer_opts.write_timeout = TIMEOUT_SECS;
+        peer_opts.idle_timeout = TIMEOUT_SECS;
+
+        // Enable HTTP/2
+        peer_opts.alpn = ALPN::H2;
+
+        let peer = Box::new(http_peer);
+        Ok(peer)
+    }
+
+    // The `upstream_request_filter` method is called before sending the request to the upstream
+    // server. We use the method to ensure that the correct headers are forwarded for gRPC
+    // requests.
+    async fn upstream_request_filter(
+        &self,
+        _session: &mut Session,
+        upstream_request: &mut RequestHeader,
+        _ctx: &mut Self::CTX,
+    ) -> Result<()>
+    where
+        Self::CTX: Send + Sync,
+    {
+        // Check if it's a gRPC request
+        if let Some(content_type) = upstream_request.headers.get("content-type") {
+            if content_type == "application/grpc" {
+                // Ensure the correct host and gRPC headers are forwarded
+                upstream_request.insert_header("content-type", "application/grpc")?;
+            }
+        }
+
+        Ok(())
+    }
+
+    // The `request_filter` method is called before processing the request.
+    // We use the method to rate limit the requests and add a unique request ID to the headers.
+    async fn request_filter(&self, session: &mut Session, _ctx: &mut Self::CTX) -> Result<bool>
+    where
+        Self::CTX: Send + Sync,
+    {
+        let client_addr = session.client_addr();
+        let user_id = client_addr.map(|addr| addr.to_string());
+
+        // Request ID is a random number
+        let request_id = rand::random::<u64>().to_string();
+        session.req_header_mut().insert_header("X-Request-ID", request_id)?;
+
+        // Retrieve the current window requests
+        let curr_window_requests = RATE_LIMITER.observe(&user_id, 1);
+
+        if curr_window_requests > MAX_REQ_PER_SEC {
+            return Self::create_too_many_requests_response(session).await;
+        };
+        Ok(false)
+    }
+
+    // If the connection fails, we retry the request [MAX_RETRIES_PER_REQUEST] times.
+    fn fail_to_connect(
+        &self,
+        _session: &mut Session,
+        _peer: &HttpPeer,
+        ctx: &mut Self::CTX,
+        mut e: Box<Error>,
+    ) -> Box<Error> {
+        if ctx.tries > MAX_RETRIES_PER_REQUEST {
+            return e;
+        }
+        ctx.tries += 1;
+        e.set_retry(true);
+        e
+    }
+
+    // The `logging` method is called after the request cycle is complete no matter the outcome.
+    // We use the method to log errors and remove the completed request from the worker queue.
+    async fn logging(&self, session: &mut Session, e: Option<&Error>, _ctx: &mut Self::CTX)
```

Comment:
nit re `///`. Also, I'd probably use the term `processed` rather than `completed` here as `completed` implies success a bit more.